### PR TITLE
core, enh: utilize pattern field to find tag location

### LINKED ISSUE #27: Handle the pattern field properly
I got an error:
```
progn: Fields not found in tags file: line
```

I generated the tags file with 
```
cd ~/var/ctags
u-ctags --pseudo-tags=+TAG_PROC_CWD  --fields=+n -n -R main parsers
```

If I removed -n from the command line and regenerated the tags file. The error was gone.

I think rejecting a tag file with -n option automatically is a simple solution.

### PR DESCRIPTION
**Pull request by AmaiKinono**
Close #27.

`citre-readtags-locate-tag` does its best to find the tag location. It makes use of the `pattern`, `line` and `name` field. It should work even if the line containing the tag is changed.

I support the `combineV2` type of pattern field directly (see <https://github.com/universal-ctags/ctags/issues/2691>). Although it has not been implemented in ctags, since my algorithm finds the nearest pattern from the line, it will cause little or no trouble if the line number shifts by one. Due to the same reason I'll not reject tags file with `combine(V1)` type of pattern field in the future, since it will work for them too.

TODO:

- [x] Fix upper components.
- [x] Decide whether we'll use this or the `line` field for `xref` integration (since `xref` uses line number forcibly, it's a little bit troublesome to integrate them).
- [x] I found not the whole line is written to the pattern field, so some tests aren't doing what they are supposed to do. I need to redesign them.
- [x] Update docstrings.

What I'd like to do after this PR:

- [ ] There are 2 `insert-file-contents` calls and they fails when the file doesn't exist. We should handle this.
- [ ] Use the kind field, maybe also `typeref`, in `citre-generate-location-str`.

~During this work I also found a bug: by using something like `--fields={line}`, we could eliminate the kind field, but Citre assumes the kind field is always the fourth field. I need to fix this.~ Fixed.

For these things, I need @masatake's help:

- We need a way to bypass `~/.ctags.d/*.ctags` when generating tags file for the tests.
- To test this function more thoroughly, I need to generate multiple tags file, with forward/backward search patterns, and `-n` option on and off. Currently there's no convenient way to do this with a single command.

Could we have a mechanism like this: we put multiple config file in `.ctags.d`, and `ctags` automatically generate a tags file for each of these? If ctags can't do this, maybe we could ask `make` for help, but either way we have to disable the user's ctags config file.

### CODE REVIEW

**Comment by masatake** [comment]
> We need a way to bypass ~/.ctags.d/*.ctags when generating tags file for the tests.

GIve "NONE" to ``--options``.  It makes warnings. You can supprss it with ``--quiet``.

```
$ ./ctags --options=NONE --quiet 
```

These options come first in the command line.

> * To test this function more thoroughly, I need to generate multiple tags file, with forward/backward search patterns, and -n option on and off. Currently there's no convenient way to do this with a single command.
>
> Could we have a mechanism like this: we put multiple config file in .ctags.d, and ctags automatically generate a tags file for each of these? If ctags can't do this, maybe we could ask make for help, but either way we have to disable the user's ctags config file.

No way to do so. I don't think adding such a feature to ctags is good.
However, this is just another use case for "edittags" command. I'm thinking about a tool for editing a tags file.
The original aim of the command is to provide a way to update "TAG_PROC_CWD" of an existing tags file.
This will be useful when moving a source tree in a local file system but you don't want to regenerate the tags file for the tree.
Such a tool can be used for converting (regenrating) tags files as you want.
But... in this case, running ctags multiple tags may be enough, I think.

**Comment by masatake** [comment]
I wonder whether using ':' in a file (or directory name) is good or bad.
In Windows(?), people use `:` in `C:\...`. Using `:` can be trouble when running the test at Windows.
(Though I don't know Windows at all.)

To avoid the case, I'm using `--` in u-ctags where I want to use `:`.

**Comment by AmaiKinono** [comment]
> In Windows(?), people use `:` in `C:\...`. Using `:` can be trouble when running the test at Windows.

Thanks. I looked it up and found `:` is an invalid character for filename in Windows. We'd better change this.

I was thinking separate the scope (core, util, tool) and the content by some symbol, but a single `-` may actually be enough if we write the scope on every test directory.

**Comment by AmaiKinono** [comment]
> GIve "NONE" to `--options`.

Thanks. Seems the right thing to do is:

```
$ ctags --options=NONE --options=.ctags.d/
```

> No way to do so. I don't think adding such a feature to ctags is good.

I agree. For now I'll create a script to generate multiple tags files. After we have enough test cases I'll think of giving them a unified interface, likely using `make` (I think it's good to update the tags file from time to time following the update of ctags itself).

File: citre.el [hunk 1]
```diff
@@ -428,29 +428,53 @@ such directory doesn't exist, nil will be returned."
...
 `citre--put-property'.
 
-What it actually does is prefix the FIELD by `citre-', and get
-that text property."
-  (get-text-property 0 (intern (concat "citre-" (symbol-name field))) str))
+When FIELD is non-nil and FROM-RECORD is nil, what it actually
+does is prefix the FIELD by `citre-', and get that text
+property.
+
+When FIELD and FROM-RECORD are both nil, it gets the record from
```
Comment (AmaiKinono) [inline citre.el:464]: Improper. Do this when FIELD is nil and FROM-RECORD is non-nil.

**Comment by AmaiKinono** [comment]
During this work I discovered several bugs. Some are found by the unit tests, some are found when I'm testing Citre on the Kernel code. I should have fix them on master and rebase this branch against master, but at the time I just quickly fixed them on this branch. I must remind myself don't do this the next time.

I think this is ready to merge, but I'd like to test it in real life for 1 or 2 days, to confirm the tools behave well for tags files with different excmd type and with the line field enabled/disabled.

Btw, the speed of Citre in the Kernel code impressed me. What a good choice to migrate to readtags!

**Pull request merged**